### PR TITLE
Mzqc conversion

### PR DESCRIPTION
**Pull request by mwalzer**
Adding first mzqc conversion (see #1)
* needs a container with py3, mzqc pronto, click (see Dockerfile)
* makes some assumptions on some metric CVs, others need to be added to psi-qc CV urgently
* probably dysfunctional nextflow process addition

### CODE REVIEW

File: elixir_proteomics_QC.nf [hunk 1]
```diff
@@ -857,8 +857,26 @@ mZML_params_for_mapping.map{
...
+    set sample_id, internal_code, checksum, file(mzidfile), file("*") from mZML_params_for_delivery.join(jsonToBeSent)
+
+    output:
+    file("${sample_id}.json")
+    
+    script:
+    """
+	python qcloud_mzqc_conversion_v0.1.py -input ${sample_id}.json -output ${sample_id}.mzqc
+    """
+}
```
Comment (mwalzer) [inline elixir_proteomics_QC.nf]: So I had some troubles to successfully execute the nf -and hence test my addition- see #5

Comment (mwalzer) [reply elixir_proteomics_QC.nf]: Pretty sure it needs some corrections

File: bin/qcloud_mzqc_conversion_v0.1.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                                )
+                            ]
+                )
+            ], 
+            analysisSoftware=[
+                mzqc.AnalysisSoftware(cvRef="MS", accession="MS:1000752", name="TOPP software", version="2.4", uri="openms.de")
+            ]
+        )
+
+    QC_9000001 = {"QC:0000029": "Total number of PSMs", "QC:0000031": "Total number of uniquely identified peptides", "QC:0000032": "Total number of identified proteins", "QC:0000007": "MS2 spectra count"}
```
Comment (mwalzer) [inline bin/qcloud_mzqc_conversion_v0.1.py]: I put them in a flexible loop-up structure as these cornerstone metrics should be in the QC cv soon.

File: bin/qcloud_mzqc_conversion_v0.1.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                                ),
+                                mzqc.CvParameter(cvRef="MS", 
+                                    accession="MS:1000031", 
+                                    name="instrument model", 
+                                    value=instrmnt
+                                )
+                            ]
+                )
+            ], 
+            analysisSoftware=[
```
Comment (mwalzer) [inline bin/qcloud_mzqc_conversion_v0.1.py]: This is a placeholder. What do we want to appear here? @rolivella

File: bin/qcloud_mzqc_conversion_v0.1.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            m = mzqc.QualityMetric()
+            m.accession = ":".join(["MS", qp['parameter']['qCCV'].split(':')[1]])
+            m.cvRef = "MS"
+            m.name = psims[m.accession].name
+            mvalue: Dict[str,List[Any]] = defaultdict(list)
+            for dp in qp['values']:
+                mvalue['QCPeptide'].append(dp['contextSource'])
+                mvalue[m.name].append(float(dp['value']))
+            m.value = mvalue
+            metrics.append(m)
```
Comment (mwalzer) [inline bin/qcloud_mzqc_conversion_v0.1.py]: Once we have the set of mzQC metrics stable and know where to get the values, the code 'complexity' drops drastically. The mzQC python lib makes it easy to produce standard files. We could even forego the json intermediate step?

File: example/qcloud_pipe_merged.mzqc [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+              "cvRef": "MS",
+              "accession": "MS:1000584",
+              "name": "mzML format"
+            },
+            "fileProperties": [
+              {
+                "cvRef": "MS",
+                "accession": "MS:1000747",
+                "name": "completion time",
+                "value": "2019-10-30 15:26:50.093306"
```
Comment (mwalzer) [inline example/qcloud_pipe_merged.mzqc]: This is wrong of course, I'd need that from the mzML. `instrument model` too.

File: example/qcloud_pipe_merged.mzqc [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+              },
+              {
+                "cvRef": "MS",
+                "accession": "MS:1000569",
+                "name": "SHA-1",
+                "value": "3f98581e6291b298c2a11a4410d7198e"
```
Comment (mwalzer) [inline example/qcloud_pipe_merged.mzqc]: BTW, which type of checksum is it you want the zip file name to contain? We'd need that for any independent verification.

File: example/qcloud_pipe_merged.mzqc [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            "name": "TOPP software",
+            "version": "2.4",
+            "uri": "openms.de"
+          }
+        ]
+      },
+      "qualityMetrics": [
+        {
+          "cvRef": "QC",
+          "accession": "QC:9000001",
```
Comment (mwalzer) [inline example/qcloud_pipe_merged.mzqc]: placeholder accession

File: example/qcloud_pipe_merged.mzqc [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+          "name": "MS2 spectra count",
+          "value": "44927"
+        },
+        {
+          "cvRef": "QC",
+          "accession": "QC:9000001",
```
Comment (mwalzer) [inline example/qcloud_pipe_merged.mzqc]: placeholder accession

File: example/qcloud_pipe_merged.mzqc [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+          "name": "Total number of PSMs",
+          "value": "21762"
+        },
+        {
+          "cvRef": "QC",
+          "accession": "QC:9000001",
```
Comment (mwalzer) [inline example/qcloud_pipe_merged.mzqc]: placeholder accession

File: example/qcloud_pipe_merged.mzqc [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+          "name": "Total number of uniquely identified peptides",
+          "value": "20051"
+        },
+        {
+          "cvRef": "QC",
+          "accession": "QC:9000001",
```
Comment (mwalzer) [inline example/qcloud_pipe_merged.mzqc]: placeholder accession

File: example/qcloud_pipe_merged.mzqc [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+          "name": "Total number of identified proteins",
+          "value": "4964"
+        },
+        {
+          "cvRef": "QC",
+          "accession": "QC:9000005",
```
Comment (mwalzer) [inline example/qcloud_pipe_merged.mzqc]: I have no clue. Just let me know, and I'll add another request for CV term entries.

File: example/qcloud_pipe_merged.mzqc [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        },
+        {
+          "cvRef": "MS",
+          "accession": "MS:1000014",
+          "name": "accuracy",
+          "value": {
+            "QCPeptide": [
+              "STLTDSLVC(Carbamidomethyl)K","SLADELALVDVLEDK","NPDDITNEEYGEFYK","LGDLYEEEMR","LAVDEEENADNNTK","FEELNMDLFR","EAALSTALSEK","RFPGYDSESK","EVSTYIK","EATTEFSVDAR","FAFQAEVNR","EQFLDGDGWTSR","YAEAVTR","TPAQFDADELR","DDVAQTDLLQIDPNFGSK"
+            ],
+            "accuracy": [
```
Comment (mwalzer) [inline example/qcloud_pipe_merged.mzqc]: So if this is the accuracy with which the identification is fitted to the MS1 feature (or reference value), then we could combine them (MS:1000014 and MS:1001844) into one metric. I am thinking something like `QCloud QC1/2 peptide detection`: QC peptide detection accuracy and intensity

**Review by rolivella — COMMENTED** [review]
Mathias, we are reviewing the code. We let you know ASAP.

File: image/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,16 @@
+FROM python:3.6.9-slim-buster
+ 
+ENV DEBIAN_FRONTEND noninteractive 
+ 
+RUN apt-get update && apt-get install -y --no-install-recommends --no-install-suggests \ 
+  git \
+  vim
```
Comment (lucacozzuto) [inline image/Dockerfile]: missing \

File: image/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+  libcurl4-openssl-dev
+
+ENV LANG C.UTF-8
+ENV LC_ALL C.UTF-8
+
+RUN echo "deb http://http.debian.net/debian buster main" > /etc/apt/sources.list.d/debian-unstable.list
+
+RUN pip install pronto pytest jupyter rpy2 biopython flask click
+RUN pip install --force-reinstall git+'https://github.com/HUPO-PSI/mzQC@mzqc-pylib#egg=MZQC-0.0.1&subdirectory=pylib'
```
Comment (lucacozzuto) [inline image/Dockerfile]: it doesn't build for me. I got the following error:

```
Step 7 : RUN pip install pronto pytest jupyter rpy2 biopython flask click
 ---> Running in 09c737dcf7e9
Collecting pronto
  Downloading https://files.pythonhosted.org/packages/01/4e/525b8fda6ed9ce8e49b9db464a13132ca82aefec5431f0f5797c521ff18f/pronto-1.1.3-py2.py3-none-any.whl (47kB)
Collecting pytest
  Downloading https://files.pythonhosted.org/packages/93/16/f6dec5178f5f4141e80dfc4812a9aba88f5f29ca881f174ab1851181d016/pytest-5.2.2-py3-none-any.whl (227kB)
Collecting jupyter
  Downloading https://files.pythonhosted.org/packages/83/df/0f5dd132200728a86190397e1ea87cd76244e42d39ec5e88efd25b2abd7e/jupyter-1.0.0-py2.py3-none-any.whl
Collecting rpy2
  Downloading https://files.pythonhosted.org/packages/d9/51/d5c41ac70e0aa3272f24cf836bd35d6246c95b98d8d811f338cec5a8098e/rpy2-3.2.2.tar.gz (162kB)
    ERROR: Command errored out with exit status 1:
     command: /usr/local/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-k_95fld0/rpy2/setup.py'"'"'; __file__='"'"'/tmp/pip-install-k_95fld0/rpy2/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-install-k_95fld0/rpy2/pip-egg-info
         cwd: /tmp/pip-install-k_95fld0/rpy2/
    Complete output (1 lines):
    There is no R_HOME and the R executable cannot be found.
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
The command '/bin/sh -c pip install pronto pytest jupyter rpy2 biopython flask click' returned a non-zero code: 1
```

Comment (lucacozzuto) [reply image/Dockerfile]: Do we really need R?
```
    There is no R_HOME and the R executable cannot be found.
`
``

**Comment by rolivella** [comment]
@mwalzer any update regarding @lucacozzuto reviews? Thank you!

File: image/Dockerfile (hunk 14 above)
Comment (mwalzer) [reply image/Dockerfile]: Yes, for the basic stuff in the conversion, we can remove the rpy2.

**Comment by lucacozzuto** [comment]
I added the feature in the master code. So I close the pull request

**Pull request closed without merging**